### ISSUE TITLE
for_each in pipeline step messes up the value

### ISSUE THREAD

**Issue by luisffc** [turbot/flowpipe#561]
I've noticed a weird behaviour on pipeline step when iterating. It seems the `each.value` always return the last element.
The example bellow shows a pipeline running 3 times, the first to run is able to create the resource named `pipeling-test-2` and the other two runs fails trying to create a resource with the same name.
It was expected to create the resources `pipeling-test-0`, `pipeling-test-1`, and `pipeling-test-2` instead.

Note that when I try the same iteration on a transform step, the names are printed correctly.

Example:

```
  step "transform" "resource_names" {
    value = [for i in range(param.amount) : "${param.role_name}-${i}"]
  }

  output "resource_names" {
    value = step.transform.resource_names.value
  }

  // // Debugging
  // step "transform" "iterating_over_resource_names" {
  //   for_each = [for name in step.transform.resource_names.value : name]
  //   value = each.value
  // }

  // // Debugging
  // output "iterating_over_resource_names" {
  //   value = step.transform.iterating_over_resource_names
  // }

  step "pipeline" "create_iam_role" {
    for_each = [for name in step.transform.resource_names.value : name]
    pipeline = aws.pipeline.create_iam_role
    args = {
      cred      = param.aws_cred
      role_name = each.value
      assume_role_policy_document = param.assume_role_policy_document
    }
  }
```

Gives me the following error:
```
$ fpr create_iam_roles --arg amount=3
[flowpipe] Execution ID: exec_cmesjr4i016p9ohpprq0
[create_iam_roles] Starting pipeline
[create_iam_roles.resource_names] Starting transform
[create_iam_roles.resource_names] Complete 1ms
[create_iam_roles.create_iam_role[0]] Starting pipeline
[create_iam_roles.create_iam_role[1]] Starting pipeline
[create_iam_roles.create_iam_role[2]] Starting pipeline
[create_iam_role] Starting pipeline
[create_iam_role] Starting pipeline
[create_iam_role] Starting pipeline
[create_iam_role.create_iam_role] Starting container
[create_iam_role.create_iam_role] Starting container
[create_iam_role.create_iam_role] Starting container
[create_iam_role.create_iam_role] Execution Error: 
An error occurred (EntityAlreadyExists) when calling the CreateRole operation: Role with name pipeling-test-2 already exists.

[create_iam_role] Output errors = [
  {
    "error": {
      "detail": "\nAn error occurred (EntityAlreadyExists) when calling the CreateRole operation: Role with name pipeling-test-2 already exists.\n",
      "instance": "fperr_cmesjski016p9ohpps3g",
      "status": 460,
      "title": "Execution Error",
      "type": "error_execution"
    },
    "pipeline": "aws.pipeline.create_iam_role",
    "pipeline_execution_id": "pexec_cmesjr4i016p9ohpprtg",
    "step": "container.create_iam_role",
    "step_execution_id": "sexec_cmesjr4i016p9ohpps20"
  }
]
[create_iam_role] Execution Error: 
An error occurred (EntityAlreadyExists) when calling the CreateRole operation: Role with name pipeling-test-2 already exists.

[create_iam_role] Failed 5s
[create_iam_roles.create_iam_role[0]] Execution Error: 
An error occurred (EntityAlreadyExists) when calling the CreateRole operation: Role with name pipeling-test-2 already exists.

[create_iam_roles.create_iam_role[0]] Failed 6s
[create_iam_role.create_iam_role] Complete 6s
[create_iam_role] Output role = {
  "Arn": "arn:aws:iam::157447638907:role/pipeling-test-2",
  "AssumeRolePolicyDocument": {
    "Statement": [
      {
        "Action": "sts:AssumeRole",
        "Effect": "Allow",
        "Principal": {
          "Service": "ec2.amazonaws.com"
        }
      }
    ],
    "Version": "2012-10-17"
  },
  "CreateDate": "2024-01-09T22:34:26+00:00",
  "Path": "/",
  "RoleId": "AROASJKE4J555JR3SLWXF",
  "RoleName": "pipeling-test-2"
}
[create_iam_role] Complete 6s
[create_iam_roles.create_iam_role[2]] Complete 6s
[create_iam_role.create_iam_role] Execution Error: 
An error occurred (EntityAlreadyExists) when calling the CreateRole operation: Role with name pipeling-test-2 already exists.

[create_iam_role] Output errors = [
  {
    "error": {
      "detail": "\nAn error occurred (EntityAlreadyExists) when calling the CreateRole operation: Role with name pipeling-test-2 already exists.\n",
      "instance": "fperr_cmesjski016p9ohpps40",
      "status": 460,
      "title": "Execution Error",
      "type": "error_execution"
    },
    "pipeline": "aws.pipeline.create_iam_role",
    "pipeline_execution_id": "pexec_cmesjr4i016p9ohpprv0",
    "step": "container.create_iam_role",
    "step_execution_id": "sexec_cmesjr4i016p9ohpps2g"
  }
]
[create_iam_role] Execution Error: 
An error occurred (EntityAlreadyExists) when calling the CreateRole operation: Role with name pipeling-test-2 already exists.

[create_iam_role] Failed 6s
[create_iam_roles.create_iam_role[1]] Execution Error: 
An error occurred (EntityAlreadyExists) when calling the CreateRole operation: Role with name pipeling-test-2 already exists.

[create_iam_roles.create_iam_role[1]] Failed 6s
[create_iam_roles] Output errors = [
  {
    "error": {
      "detail": "\nAn error occurred (EntityAlreadyExists) when calling the CreateRole operation: Role with name pipeling-test-2 already exists.\n",
      "instance": "fperr_cmesjski016p9ohpps3g",
      "status": 460,
      "title": "Execution Error",
      "type": "error_execution"
    },
    "pipeline": "aws.pipeline.create_iam_role",
    "pipeline_execution_id": "pexec_cmesjr4i016p9ohpprtg",
    "step": "container.create_iam_role",
    "step_execution_id": "sexec_cmesjr4i016p9ohpps20"
  },
  {
    "error": {
      "detail": "\nAn error occurred (EntityAlreadyExists) when calling the CreateRole operation: Role with name pipeling-test-2 already exists.\n",
      "instance": "fperr_cmesjski016p9ohpps40",
      "status": 460,
      "title": "Execution Error",
      "type": "error_execution"
    },
    "pipeline": "aws.pipeline.create_iam_role",
    "pipeline_execution_id": "pexec_cmesjr4i016p9ohpprv0",
    "step": "container.create_iam_role",
    "step_execution_id": "sexec_cmesjr4i016p9ohpps2g"
  }
]
[create_iam_roles] Execution Error: 
An error occurred (EntityAlreadyExists) when calling the CreateRole operation: Role with name pipeling-test-2 already exists.

[create_iam_roles] Execution Error: 
An error occurred (EntityAlreadyExists) when calling the CreateRole operation: Role with name pipeling-test-2 already exists.

[create_iam_roles] Failed with 2 errors 6s
```

If I comment out the pipeline step and uncomment the debugging steps, I get the following output:
```
$ fpr create_iam_roles --arg amount=3
[flowpipe] Execution ID: exec_cmeskaci016pb3idil60
[create_iam_roles] Starting pipeline
[create_iam_roles.resource_names] Starting transform
[create_iam_roles.resource_names] Complete 2ms
[create_iam_roles.iterating_over_resource_names[0]] Starting transform
[create_iam_roles.iterating_over_resource_names[0]] Complete 1ms
[create_iam_roles.iterating_over_resource_names[1]] Starting transform
[create_iam_roles.iterating_over_resource_names[1]] Complete 3ms
[create_iam_roles.iterating_over_resource_names[2]] Starting transform
[create_iam_roles.iterating_over_resource_names[2]] Complete 1ms
[create_iam_roles] Output iterating_over_resource_names = {
  "0": {
    "value": "pipeling-test-0"
  },
  "1": {
    "value": "pipeling-test-1"
  },
  "2": {
    "value": "pipeling-test-2"
  }
}
[create_iam_roles] Output resource_names = [
  "pipeling-test-0",
  "pipeling-test-1",
  "pipeling-test-2"
]
[create_iam_roles] Complete 23ms exec_cmeskaci016pb3idil60
```

**Comment by luisffc** [comment]
`fpr` is an alias for `flowpipe pipeline run`

**Comment by vhadianto** [comment]
@luisffc I'm unable to reproduce this. I have this sample pipeline that I think a simplification of your pipeline (with no external access required)

```hcl
pipeline "parent_with_foreach" {
    param "times" {
        type = number
        default = 3
    }

    step "transform" "names" {
        value = [for i in range(param.times) : "name-${i}"]
    }

    step "pipeline" "nested_one" {
        for_each = [for name in step.transform.names.value : name]
        pipeline = pipeline.nested_one
        args = {
            name = each.value
        }
    }

    output "vals" {
        value = step.pipeline.nested_one
    }
}

pipeline "nested_one" {
    param "name" {
        type = string
        default = "default value"
    }
    step "transform" "name" {
        value = param.name
    }
    output "val" {
        value = step.transform.name.value
    }
}
```

The result is as expected, I see 3 pipeline execution with param: `name-0`, `name-1` and `name-2`.

Can you please reproduce in a simple pipeline as above so I can have a look?

**Comment by luisffc** [comment]
@vhadianto The issue is on container step.

I've changed the sample pipeline to use echo in a container and noticed the stdout is always the same. It seems to be caching the value somewhere:

```pipeline "parent_with_foreach" {
    param "times" {
        type = number
        default = 3
    }

    step "transform" "names" {
        value = [for i in range(param.times) : "name-${i}"]
    }

    step "pipeline" "nested_one" {
        for_each = step.transform.names.value
        pipeline = pipeline.nested_one
        args = {
            name = each.value
        }
    }

    output "vals" {
        value = step.pipeline.nested_one
    }
}

pipeline "nested_one" {
    param "name" {
        type = string
        default = "default value"
    }
    step "container" "name" {
      image = "bash"
      cmd = ["echo", param.name]
    }
    output "val" {
        value = step.container.name
    }
}
```

output
```
[flowpipe] Execution ID: exec_cmff9rsi016uuh1ap880
[parent_with_foreach] Starting pipeline
[parent_with_foreach.names] Starting transform
[parent_with_foreach.names] Complete 2ms
[parent_with_foreach.nested_one[1]] Starting pipeline
[parent_with_foreach.nested_one[0]] Starting pipeline
[parent_with_foreach.nested_one[2]] Starting pipeline
[nested_one] Starting pipeline
[nested_one] Starting pipeline
[nested_one] Starting pipeline
[nested_one.name] Starting container
[nested_one.name] Starting container
[nested_one.name] Starting container
[nested_one.name] Complete 2s
[nested_one] Output val = {
  "cmd": [
    "echo",
    "name-2"
  ],
  "container_id": "REDACTEDfb5f3178f163c75ad310ad8f",
  "exit_code": 0,
  "image": "bash",
  "lines": [
    null
  ],
  "name": "name",
  "stderr": "",
  "stdout": "name-2\n"
}
[nested_one] Complete 2s
[parent_with_foreach.nested_one[2]] Complete 2s
[nested_one.name] Complete 2s
[nested_one] Output val = {
  "cmd": [
    "echo",
    "name-1"
  ],
  "container_id": "REDACTED12ea5067c29a8e11903b08cb",
  "exit_code": 0,
  "image": "bash",
  "lines": [
    null
  ],
  "name": "name",
  "stderr": "",
  "stdout": "name-2\n"
}
[nested_one] Complete 2s
[parent_with_foreach.nested_one[1]] Complete 2s
[nested_one.name] Complete 2s
[nested_one] Output val = {
  "cmd": [
    "echo",
    "name-0"
  ],
  "container_id": "REDACTED423e0b0d31ac730bb57ccc4b",
  "exit_code": 0,
  "image": "bash",
  "lines": [
    null
  ],
  "name": "name",
  "stderr": "",
  "stdout": "name-2\n"
}
[nested_one] Complete 2s
[parent_with_foreach.nested_one[0]] Complete 2s
[parent_with_foreach] Output vals = {
  "0": {
    "args": {
      "name": "name-0"
    },
    "output": {
      "val": {
        "cmd": [
          "echo",
          "name-0"
        ],
        "container_id": "REDACTED423e0b0d31ac730bb57ccc4b",
        "exit_code": 0,
        "image": "bash",
        "lines": [
          null
        ],
        "name": "name",
        "stderr": "",
        "stdout": "name-2\n"
      }
    },
    "pipeline": "bulk_resource_creator.pipeline.nested_one"
  },
  "1": {
    "args": {
      "name": "name-1"
    },
    "output": {
      "val": {
        "cmd": [
          "echo",
          "name-1"
        ],
        "container_id": "REDACTED12ea5067c29a8e11903b08cb",
        "exit_code": 0,
        "image": "bash",
        "lines": [
          null
        ],
        "name": "name",
        "stderr": "",
        "stdout": "name-2\n"
      }
    },
    "pipeline": "bulk_resource_creator.pipeline.nested_one"
  },
  "2": {
    "args": {
      "name": "name-2"
    },
    "output": {
      "val": {
        "cmd": [
          "echo",
          "name-2"
        ],
        "container_id": "REDACTEDfb5f3178f163c75ad310ad8f",
        "exit_code": 0,
        "image": "bash",
        "lines": [
          null
        ],
        "name": "name",
        "stderr": "",
        "stdout": "name-2\n"
      }
    },
    "pipeline": "bulk_resource_creator.pipeline.nested_one"
  }
}
[parent_with_foreach] Complete 2s
```

**Comment by vhadianto** [comment]
I see. @luisffc can you please create the following test:
* self contained
* has no external SaaS query with authentication
* use small container (Alpine)

As it stands I can't run the above pipeline fragment. Will be good to get the issue from the source.

**Comment by luisffc** [comment]
oh, it was a typo. I skipped the line break on the first line:
` ```pipeline "parent_with_foreach" { `
Just edited

**Comment by graza-io** [comment]
Fixed in v0.2.1
